File: tools/dogfood_coverage_test.py
#!/usr/bin/env python3
"""Tests for tools/dogfood_coverage.py."""
from __future__ import annotations

import importlib.util
import sys
import tempfile
import unittest
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SCRIPT = ROOT / "tools" / "dogfood_coverage.py"


def load():
    sys.path.insert(0, str(SCRIPT.parent))
    spec = importlib.util.spec_from_file_location("dogfood_coverage", SCRIPT)
    assert spec and spec.loader
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


class DogfoodCoverageTest(unittest.TestCase):
    def test_evaluate_on_real_repo_has_schema_and_hard_kpis(self) -> None:
        mod = load()
        payload = mod.evaluate(ROOT, env={})  # empty env => guard default ON
        self.assertEqual(payload["schema"], "dogfood-coverage/1")
        self.assertGreaterEqual(payload["coverage"], 0.0)
        self.assertLessEqual(payload["coverage"], 100.0)
        self.assertIsInstance(payload["dogfood_debt"], int)
        self.assertIn(payload["grade"], {"A", "B", "C", "D", "F"})
        keys = {k["key"] for k in payload["kpis"]}
        for expected in ("fleet_leaf_guarded", "fak_bin_resolvable", "guard_default_on",
                         "issue_dispatch_wired", "guard_verb_present"):
            self.assertIn(expected, keys)

    def test_guard_disabled_env_lowers_coverage_and_raises_debt(self) -> None:
        mod = load()
        on = mod.evaluate(ROOT, env={})
        off = mod.evaluate(ROOT, env={"FLEET_DOGFOOD_GUARD": "0"})
        # Turning the dogfood OFF must FAIL the live-env KPI (and at least the leaf
        # guard KPI), so debt rises and coverage drops vs the default-on run.
        self.assertGreater(off["dogfood_debt"], on["dogfood_debt"])
        self.assertLess(off["coverage"], on["coverage"])
        off_keys = {k["key"]: k["ok"] for k in off["kpis"]}
        self.assertFalse(off_keys["guard_default_on"])

    def test_count_audit_rows_counts_nonblank_jsonl_lines(self) -> None:
        mod = load()
        with tempfile.TemporaryDirectory() as td:
            root = Path(td)
            jdir = root / ".dispatch-runs" / "guard-audit"
            jdir.mkdir(parents=True)
            (jdir / "gateway-claude.jsonl").write_text(
                '{"seq":1}\n{"seq":2}\n\n{"seq":3}\n', encoding="utf-8")
            (jdir / "docs-claude.jsonl").write_text('{"seq":1}\n', encoding="utf-8")
            rows, journals = mod.count_audit_rows(root)
        self.assertEqual(rows, 4)       # 3 + 1 non-blank lines
        self.assertEqual(journals, 2)

    def test_diagnose_audit_gap_distinguishes_the_three_blanks(self) -> None:
        mod = load()
        with tempfile.TemporaryDirectory() as td:
            root = Path(td)
            # (1) no journal dir at all -> "no guarded worker has run"
            msg_no_dir = mod.diagnose_audit_gap(root)
            self.assertIn("no guarded worker", msg_no_dir)

            jdir = root / ".dispatch-runs" / "guard-audit"
            jdir.mkdir(parents=True)
            # (2) dir but no files -> "configured but never exercised"
            msg_empty_dir = mod.diagnose_audit_gap(root)
            self.assertIn("never exercised", msg_empty_dir)

            # (3) dir + only-blank files -> the silent empty-turn signature
            (jdir / "docs-claude.jsonl").write_text("\n\n", encoding="utf-8")
            msg_blank_files = mod.diagnose_audit_gap(root)
            self.assertIn("all blank", msg_blank_files)
            self.assertIn("auth/login", msg_blank_files)

            # The three blanks must be distinguishable (the whole point).
            self.assertEqual(len({msg_no_dir, msg_empty_dir, msg_blank_files}), 3)

        # And when rows exist, there is no gap to explain.
        with tempfile.TemporaryDirectory() as td:
            root = Path(td)
            jdir = root / ".dispatch-runs" / "guard-audit"
            jdir.mkdir(parents=True)
            (jdir / "g.jsonl").write_text('{"seq":1}\n', encoding="utf-8")
            # diagnose is only consulted when rows==0; here it would report the
            # blank-files branch is NOT hit because the file has a row.
            self.assertEqual(mod.count_audit_rows(root)[0], 1)

    def test_grade_ladder(self) -> None:
        mod = load()
        self.assertEqual(mod._grade(100.0, 0), "A")
        self.assertEqual(mod._grade(70.0, 0), "B")
        self.assertEqual(mod._grade(70.0, 1), "C")
        self.assertEqual(mod._grade(50.0, 2), "D")
        self.assertEqual(mod._grade(10.0, 5), "F")

    def test_check_mode_returns_zero_when_no_hard_debt(self) -> None:
        mod = load()
        # On this repo the HARD KPIs pass (fak is built), so --check should exit 0.
        rc = mod.main(["--check"])
        self.assertIn(rc, (0, 1))   # tolerate a host without fak built (fail-open)

    def test_render_is_stringy_and_mentions_coverage(self) -> None:
        mod = load()
        payload = mod.evaluate(ROOT, env={})
        out = mod.render(payload)
        self.assertIn("dogfood-coverage", out)
        self.assertIn("grade", out)


if __name__ == "__main__":
    unittest.main()

File: internal/dogfoodcoverage/dogfoodcoverage_test.go
package dogfoodcoverage

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func findTestRepoRoot(t *testing.T) string {
	t.Helper()
	wd, err := os.Getwd()
	if err != nil {
		t.Fatalf("getwd: %v", err)
	}
	dir := wd
	for {
		if _, err := os.Stat(filepath.Join(dir, "go.mod")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			t.Fatalf("cannot find repo root from %s", wd)
		}
		dir = parent
	}
}

func TestEvaluateOnRealRepoHasSchemaAndHardKPIs(t *testing.T) {
	root := findTestRepoRoot(t)
	payload := Evaluate(root, map[string]string{})

	if payload.Schema != Schema {
		t.Errorf("schema = %q, want %q", payload.Schema, Schema)
	}
	if payload.Coverage < 0.0 || payload.Coverage > 100.0 {
		t.Errorf("coverage = %f, want [0.0, 100.0]", payload.Coverage)
	}
	switch payload.Grade {
	case "A", "B", "C", "D", "F":
	default:
		t.Errorf("unexpected grade %q", payload.Grade)
	}

	keys := make(map[string]bool)
	for _, k := range payload.KPIs {
		keys[k.Key] = true
	}

	expectedHard := []string{
		"fleet_leaf_guarded",
		"fak_bin_resolvable",
		"guard_default_on",
		"issue_dispatch_wired",
		"guard_verb_present",
	}
	for _, exp := range expectedHard {
		if !keys[exp] {
			t.Errorf("missing hard KPI key %q", exp)
		}
	}
}

func TestGuardDisabledEnvLowersCoverageAndRaisesDebt(t *testing.T) {
	root := findTestRepoRoot(t)
	on := Evaluate(root, map[string]string{})
	off := Evaluate(root, map[string]string{"FLEET_DOGFOOD_GUARD": "0"})

	if off.DogfoodDebt <= on.DogfoodDebt {
		t.Errorf("off.DogfoodDebt (%d) should be > on.DogfoodDebt (%d)", off.DogfoodDebt, on.DogfoodDebt)
	}
	if off.Coverage >= on.Coverage {
		t.Errorf("off.Coverage (%f) should be < on.Coverage (%f)", off.Coverage, on.Coverage)
	}

	offKeys := make(map[string]bool)
	for _, k := range off.KPIs {
		offKeys[k.Key] = k.OK
	}
	if offKeys["guard_default_on"] {
		t.Errorf("guard_default_on should be false when FLEET_DOGFOOD_GUARD=0")
	}
}

func TestEvaluateDegradesWhenDispatchWorkerMissing(t *testing.T) {
	td := t.TempDir()
	if err := os.Mkdir(filepath.Join(td, "tools"), 0755); err != nil {
		t.Fatalf("mkdir tools: %v", err)
	}
	payload := Evaluate(td, map[string]string{})

	kpis := make(map[string]KPI)
	for _, k := range payload.KPIs {
		kpis[k.Key] = k
	}

	for _, key := range []string{"fleet_leaf_guarded", "fak_bin_resolvable", "guard_default_on"} {
		k, ok := kpis[key]
		if !ok {
			t.Fatalf("missing KPI %q", key)
		}
		if k.OK {
			t.Errorf("%s: OK should be false", key)
		}
		if !k.Hard {
			t.Errorf("%s: Hard should be true", key)
		}
		if !strings.Contains(k.Evidence, "dispatch surface moved") {
			t.Errorf("%s: Evidence %q should contain 'dispatch surface moved'", key, k.Evidence)
		}
	}
}

func TestEvaluateDegradesWhenDispatchWorkerSymbolRenamed(t *testing.T) {
	td := t.TempDir()
	toolsDir := filepath.Join(td, "tools")
	if err := os.Mkdir(toolsDir, 0755); err != nil {
		t.Fatalf("mkdir tools: %v", err)
	}
	dwPath := filepath.Join(toolsDir, "dispatch_worker.py")
	if err := os.WriteFile(dwPath, []byte("def build_command(*a, **k):\n    return ['claude']\n"), 0644); err != nil {
		t.Fatalf("write dispatch_worker.py: %v", err)
	}

	payload := Evaluate(td, map[string]string{})

	kpis := make(map[string]KPI)
	for _, k := range payload.KPIs {
		kpis[k.Key] = k
	}

	for _, key := range []string{"fleet_leaf_guarded", "fak_bin_resolvable", "guard_default_on"} {
		k, ok := kpis[key]
		if !ok {
			t.Fatalf("missing KPI %q", key)
		}
		if k.OK {
			t.Errorf("%s: OK should be false", key)
		}
		if !k.Hard {
			t.Errorf("%s: Hard should be true", key)
		}
		if !strings.Contains(k.Evidence, "dispatch surface moved") {
			t.Errorf("%s: Evidence %q should contain 'dispatch surface moved'", key, k.Evidence)
		}
	}
}

func TestCountAuditRowsCountsNonblankJSONLLines(t *testing.T) {
	td := t.TempDir()
	cfg := t.TempDir()
	jdir := filepath.Join(td, ".dispatch-runs", "guard-audit")
	if err := os.MkdirAll(jdir, 0755); err != nil {
		t.Fatalf("mkdir guard-audit: %v", err)
	}
	if err := os.WriteFile(filepath.Join(jdir, "gateway-claude.jsonl"), []byte("{\"seq\":1}\n{\"seq\":2}\n\n{\"seq\":3}\n"), 0644); err != nil {
		t.Fatalf("write gateway-claude.jsonl: %v", err)
	}
	if err := os.WriteFile(filepath.Join(jdir, "docs-claude.jsonl"), []byte("{\"seq\":1}\n"), 0644); err != nil {
		t.Fatalf("write docs-claude.jsonl: %v", err)
	}

	rows, journals := CountAuditRows(td, map[string]string{"XDG_CONFIG_HOME": cfg})
	if rows != 4 {
		t.Errorf("rows = %d, want 4", rows)
	}
	if journals != 2 {
		t.Errorf("journals = %d, want 2", journals)
	}
}

func TestDiagnoseAuditGapDistinguishesTheThreeBlanks(t *testing.T) {
	td := t.TempDir()

	// (1) no journal dir at all
	msgNoDir := DiagnoseAuditGap(td)
	if !strings.Contains(msgNoDir, "no guarded worker") {
		t.Errorf("msgNoDir = %q, want 'no guarded worker'", msgNoDir)
	}

	// (2) dir but no files
	jdir := filepath.Join(td, ".dispatch-runs", "guard-audit")
	if err := os.MkdirAll(jdir, 0755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	msgEmptyDir := DiagnoseAuditGap(td)
	if !strings.Contains(msgEmptyDir, "never exercised") {
		t.Errorf("msgEmptyDir = %q, want 'never exercised'", msgEmptyDir)
	}

	// (3) dir + only-blank files
	if err := os.WriteFile(filepath.Join(jdir, "docs-claude.jsonl"), []byte("\n\n"), 0644); err != nil {
		t.Fatalf("write: %v", err)
	}
	msgBlankFiles := DiagnoseAuditGap(td)
	if !strings.Contains(msgBlankFiles, "all blank") {
		t.Errorf("msgBlankFiles = %q, want 'all blank'", msgBlankFiles)
	}
	if !strings.Contains(msgBlankFiles, "auth/login") {
		t.Errorf("msgBlankFiles = %q, want 'auth/login'", msgBlankFiles)
	}

	// Distinct messages
	set := map[string]bool{
		msgNoDir:      true,
		msgEmptyDir:   true,
		msgBlankFiles: true,
	}
	if len(set) != 3 {
		t.Errorf("expected 3 distinct diagnostic messages, got %d", len(set))
	}

	// When rows exist, there is no gap
	if err := os.WriteFile(filepath.Join(jdir, "g.jsonl"), []byte("{\"seq\":1}\n"), 0644); err != nil {
		t.Fatalf("write g.jsonl: %v", err)
	}
	rows, _ := CountAuditRows(td, nil)
	if rows != 1 {
		t.Errorf("rows = %d, want 1", rows)
	}
	msgWithRows := DiagnoseAuditGap(td)
	if msgWithRows != "" {
		t.Errorf("DiagnoseAuditGap when rows exist = %q, want empty", msgWithRows)
	}
}

func TestGradeLadder(t *testing.T) {
	cases := []struct {
		coverage float64
		debt     int
		want     string
	}{
		{100.0, 0, "A"},
		{95.0, 0, "A"},
		{70.0, 0, "B"},
		{70.0, 1, "C"},
		{50.0, 2, "D"},
		{10.0, 5, "F"},
	}
	for _, tc := range cases {
		got := Grade(tc.coverage, tc.debt)
		if got != tc.want {
			t.Errorf("Grade(%.1f, %d) = %q, want %q", tc.coverage, tc.debt, got, tc.want)
		}
	}
}

func TestFormatReport(t *testing.T) {
	root := findTestRepoRoot(t)
	payload := Evaluate(root, map[string]string{})
	out := FormatReport(payload)
	if !strings.Contains(out, "dogfood-coverage") {
		t.Errorf("output missing 'dogfood-coverage': %s", out)
	}
	if !strings.Contains(out, "grade") {
		t.Errorf("output missing 'grade': %s", out)
	}
}

func TestCountAuditRowsWithUserJournal(t *testing.T) {
	td := t.TempDir()
	cfg := t.TempDir()
	// Must have .git present to activate user journal
	if err := os.WriteFile(filepath.Join(td, ".git"), []byte("gitdir: dummy"), 0644); err != nil {
		t.Fatalf("write .git: %v", err)
	}
	userDir := filepath.Join(cfg, "fak")
	if err := os.MkdirAll(userDir, 0755); err != nil {
		t.Fatalf("mkdir userDir: %v", err)
	}
	if err := os.WriteFile(filepath.Join(userDir, "guard-audit.jsonl"), []byte("{\"u\":1}\n{\"u\":2}\n"), 0644); err != nil {
		t.Fatalf("write user journal: %v", err)
	}

	rows, journals := CountAuditRows(td, map[string]string{"XDG_CONFIG_HOME": cfg})
	if rows != 2 {
		t.Errorf("rows = %d, want 2", rows)
	}
	if journals != 1 {
		t.Errorf("journals = %d, want 1", journals)
	}
}

func TestCheckGateBehavior(t *testing.T) {
	root := findTestRepoRoot(t)
	// Guard on => debt 0 => check passes
	repOn := Evaluate(root, map[string]string{})
	if repOn.DogfoodDebt > 0 {
		t.Fatalf("expected debt 0 when guard is on, got %d", repOn.DogfoodDebt)
	}
	// Guard off => debt > 0 => check fails
	repOff := Evaluate(root, map[string]string{"FLEET_DOGFOOD_GUARD": "0"})
	if repOff.DogfoodDebt == 0 {
		t.Fatalf("expected debt > 0 when FLEET_DOGFOOD_GUARD=0, got 0")
	}
}
